File: src/Source/SuiviCommande.java
package source;

import org.json.*;
import java.util.Scanner;
import java.io.*;

public class SuiviCommande implements Serializable
{
    private int idClient;
    private int idRevendeur;
    private int idTare;
    private int numeroDeCommande;
    private int numeroDeLot;
    private String typeEnergie;
    private String modeExtraction;
    private int quantiteDemander;
    private int quantiteEnvoyer;
    private int prixUnites;
    private int prixTotal;

    public SuiviCommande(int idClient,int idRevendeur,int idTare,int numeroDeCommande,int numeroDeLot, String typeEnergie,String modeExtraction,int quantiteDemander,int quantiteEnvoyer,int prixUnites,int prixTotal) {
        this.idClient = idClient;
        this.idRevendeur = idRevendeur;
        this.idTare = idTare;
        this.numeroDeCommande = numeroDeCommande;
        this.numeroDeLot = numeroDeLot;
        this.typeEnergie = typeEnergie; 
        this.modeExtraction = modeExtraction;   
        this.quantiteDemander = quantiteDemander;
        this.quantiteEnvoyer = quantiteEnvoyer;
        this.prixUnites = prixUnites;
        this.prixTotal = prixTotal;
    }

    public static String[] Split(String reference)
    {
        String[] partiRef = reference.split("-");
        return partiRef;
    }

    public int getIdClient() {
        return idClient;
    }
    public int getIdRevendeur() {
        return idRevendeur;
    }
    public int getIdTare() {
        return idTare;
    }
    public String getModeExtraction() {
        return modeExtraction;
    }
    public int getNumeroDeCommande() {
        return numeroDeCommande;
    }
    public int getPrixTotal() {
        return prixTotal;
    }
    public int getPrixUnites() {
        return prixUnites;
    }
    public int getQuantiteDemander() {
        return quantiteDemander;
    }
    public int getQuantiteEnvoyer() {
        return quantiteEnvoyer;
    }
    public String getTypeEnergie() {
        return typeEnergie;
    }
    public int getNumeroDeLot() 
    {
        return numeroDeLot;
    }

    public JSONObject toJson()
    {
        JSONObject json = new JSONObject();
        json.put("IdClient", this.idClient);
        json.put("IdRevendeur", this.idRevendeur);
        json.put("IdTare", this.idTare);
        json.put("NumeroDeCommande", this.numeroDeCommande);
        json.put("NumeroDeLot", this.numeroDeLot);
        json.put("TypeEnergie", this.typeEnergie);
        json.put("ModeExtraction", this.modeExtraction);
        json.put("QuantiteDemander", this.quantiteDemander);
        json.put("QuantiteEnvoyer", this.quantiteEnvoyer);
        json.put("PrixUnites", this.prixUnites);
        json.put("PrixTotal", this.prixTotal);

        return json;
    }

    public static SuiviCommande FromJSON(String json)
    {
        JSONObject objet = new JSONObject(json);

        int IdClient = objet.getInt("IdClient");
        int IdRevendeur = objet.getInt("IdRevendeur");
        int IdTare = objet.getInt("IdTare");
        int numeroDeCommande = objet.getInt("NumeroDeCommande");
        int numeroDeLot = objet.getInt("NumeroDeLot");
        String typeEnergie = objet.getString("TypeEnergie");
        String modeExtraction = objet.getString("ModeExtraction");
        int quantiteDemander = objet.getInt("QuantiteDemander");
        int quantiteEnvoyer = objet.getInt("QuantiteEnvoyer");
        int prixUnites = objet.getInt("PrixUnites");
        int prixTotal = objet.getInt("PrixTotal");

        SuiviCommande suiviCommande = new SuiviCommande(
                            IdClient, 
                            IdRevendeur, 
                            IdTare, 
                            numeroDeCommande, 
                            numeroDeLot,
                            typeEnergie, 
                            modeExtraction, 
                            quantiteDemander, 
                            quantiteEnvoyer, 
                            prixUnites,
                            prixTotal
                            );
        return suiviCommande;
    }

    @Override
    public String toString() {
            return (                                                                              "\n" +
                                "  => Id Client : "              + this.idClient                + "\n" +
                                "  => Id Revendeur : "           + this.idRevendeur             + "\n" +
                                "  => Id Tare : "                + this.idTare                  + "\n" +
                                "  => Numero de commande : "     + this.numeroDeCommande        + "\n" +
                                "  => Numero de lot : "          + this.numeroDeLot            + "\n" +
                                "  => Type Energie : "           + this.typeEnergie             + "\n" +
                                "  => Mode Extraction : "        + this.modeExtraction          + "\n" +
                                "  => Quantite Demander : "      + this.quantiteDemander        + "\n" +
                                "  => Quantite Envoyer : "       + this.quantiteEnvoyer         + "\n" +
                                "  => Prix Units : "             + this.prixUnites              + "\n" + 
                                "  => Prix Totals : "            + this.prixTotal               + "\n"
                    );
    }

}

File: src/revendeur/SuiviCommande.php
<?php

    #author Romain Cogné (romain.cogne@etudiant.univ-reims.fr)
    #author Quentin Juilliard (quentin.juilliard@etudiant.univ-reims.fr)

    class SuiviCommande implements JsonSerializable{
        //Variables
        private int $idClient;
        private int $idRevendeur;
        private int $idTare;
        private int $numeroDeCommande;
        private int $numerodDeLot;
        private String $typeEnergie;
        private String $modeExtraction;
        private int $quantiteDemander;
        private int $quantiteEnvoyer;
        private int $prixUnites;
        private int $prixTotal;

        //Constructeur
        public function __construct($idClient,$idRevendeur,$idTare,$numeroDeCommande,$numerodDeLot,$typeEnergie,$modeExtraction,$quantiteDemander,$quantiteEnvoyer,$prixUnites,$prixTotal)
        {    
            $this->idClient= $idClient;
            $this->idRevendeur= $idRevendeur;
            $this->idTare= $idTare;
            $this->numeroDeCommande= $numeroDeCommande;
            $this->numerodDeLot= $numerodDeLot;
            $this->typeEnergie= $typeEnergie;
            $this->modeExtraction= $modeExtraction;
            $this->quantiteDemander= $quantiteDemander;
            $this->quantiteEnvoyer= $quantiteEnvoyer;
            $this->prixUnites= $prixUnites;
            $this->prixTotal= $prixTotal;
            
        }

        public function getIdClient()
        {
            return $this->idClient;
        }
        public function getIdRevendeur()
        {
            return $this->idRevendeur;
        }
        public function getIdTare()
        {
            return $this->idTare;
        }
        public function getNumeroDeCommande()
        {
            return $this->numeroDeCommande;
        }
        public function getNumerodDeLot()
        {
            return $this->numerodDeLot;
        }
        public function getTypeEnergie()
        {
            return $this->typeEnergie;
        }
        public function getModeExtraction()
        {
            return $this->modeExtraction;
        }
        public function getQuantiteDemander()
        {
            return $this->quantiteDemander;
        }
        public function getQuantiteEnvoyer()
        {
            return $this->quantiteEnvoyer;
        }
        public function getPrixUnites()
        {
            return $this->prixUnites;
        }
        public function getPrixTotal()
        {
            return $this->prixTotal;
        }


        //Pour sérialiser l'objet
        public function jsonSerialize() : array 
        {
            return [
                "IdClient"=>$this->idClient,
                "IdRevendeur"=>$this->idRevendeur,
                "IdTare"=>$this->idTare,
                "NumeroDeCommande"=>$this->numeroDeCommande,
                "NumeroDeLot"=>$this->numerodDeLot,
                "TypeEnergie"=>$this->typeEnergie,
                "ModeExtraction"=>$this->modeExtraction,
                "QuantiteDemander"=>$this->quantiteDemander,
                "QuantiteEnvoyer"=>$this->quantiteEnvoyer,
                "PrixUnites"=>$this->prixUnites,
                "PrixTotal"=>$this->prixTotal,
            ];
        }

        // decode un Json 
        public static function fromJson(array $arr) : SuiviCommande 
        {   
            $idClient = $arr['IdClient'];
            $idRevendeur = $arr['IdRevendeur'];
            $idTare = $arr['IdTare'];
            $numeroDeCommande = $arr['NumeroDeCommande'];
            $numerodDeLot = $arr['NumeroDeLot'];
            $typeEnergie = $arr['TypeEnergie'];
            $modeExtraction = $arr['ModeExtraction'];
            $quantiteDemander = $arr['QuantiteDemander'];
            $quantiteEnvoyer = $arr['QuantiteEnvoyer'];
            $prixUnites = $arr['PrixUnites'];
            $prixTotal = $arr['PrixTotal'];

            return new SuiviCommande (
                                    $idClient,
                                    $idRevendeur,
                                    $idTare,
                                    $numeroDeCommande,
                                    $numerodDeLot,
                                    $typeEnergie,
                                    $modeExtraction,
                                    $quantiteDemander,
                                    $quantiteEnvoyer,
                                    $prixUnites,
                                    $prixTotal,
                                );
        }

        //Affcihe la commande
        public function __toString()
        {
            return  " -> IdClient : " . $this->idClient . "<br/>" . 
                    " -> IdRevendeur : " . $this->idRevendeur . "<br/>" . 
                    " -> IdTare : " . $this->idTare . "<br/>" . 
                    " -> numeroDeCommande : " . $this->numeroDeCommande . "<br/>".
                    " -> numerodDeLot : " . $this->numerodDeLot . "<br/>".
                    " -> typeEnergie : " . $this->typeEnergie . "<br/>".
                    " -> modeExtraction : " . $this->modeExtraction . "<br/>".
                    " -> quantiteDemander : " . $this->quantiteDemander . "<br/>".
                    " -> quantiteEnvoyer : " . $this->quantiteEnvoyer . "<br/>".
                    " -> prixUnites : " . $this->prixUnites . "<br/>".
                    " -> prixTotal : " . $this->prixTotal . "<br/>"
                    ;
        }

        //Créattion de la socket d'envoie
        public function envoieSocket(){

            switch ($this->typeEnergie) 
            {
                case 'Electricite':
                    $port = 1021;
                    break;

                case 'Gaz':
                    $port = 1022;
                    break;

                case 'Petrole':
                    $port = 1023;
                    break;
                
                default :
                    break;
            }

            $data = urlencode(json_encode($this));
            
            $options = [
                'http' =>
                [
                    'method'  => 'POST',
                    'header'  => 'Content-type: application/x-www-form-urlencoded',
                    'content' => $data
                ]
            ];
            // Envoi de la requête et lecture du JSON reçu
            $URL = "http://localhost:".$port."/traitement";
            //echo $URL . "<br/>";
            $contexte  = stream_context_create($options);
            $jsonTexte = @file_get_contents($URL, false, $contexte);
            //echo "JSON TEXTE => " . $jsonTexte;

            return $jsonTexte;
        }
        
    }
?>
